pkg/prometheus/promcfg.go

```diff
@@ -431,170 +438,91 @@ func (cg *ConfigGenerator) Generate(
+	globalItems = cg.AppendQueryLogFile(globalItems, queryLogFile)
 	cfg = append(cfg, yaml.MapItem{Key: "global", Value: globalItems})
 
-	if p.Spec.RuleSelector != nil {
-		ruleFilePaths := []string{}
-		for _, name := range ruleConfigMapNames {
-			ruleFilePaths = append(ruleFilePaths, rulesDir+"/"+name+"/*.yaml")
-		}
-		cfg = append(cfg, yaml.MapItem{
-			Key:   "rule_files",
-			Value: ruleFilePaths,
-		})
-	}
-
-	sMonIdentifiers := make([]string, len(sMons))
-	i := 0
-	for k := range sMons {
-		sMonIdentifiers[i] = k
-		i++
-	}
-
-	// Sorting ensures, that we always generate the config in the same order.
-	sort.Strings(sMonIdentifiers)
-
-	pMonIdentifiers := make([]string, len(pMons))
-	i = 0
-	for k := range pMons {
-		pMonIdentifiers[i] = k
-		i++
-	}
-
-	// Sorting ensures, that we always generate the config in the same order.
-	sort.Strings(pMonIdentifiers)
-
-	probeIdentifiers := make([]string, len(probes))
-	i = 0
-	for k := range probes {
-		probeIdentifiers[i] = k
-		i++
-	}
-
-	// Sorting ensures, that we always generate the config in the same order.
-	sort.Strings(probeIdentifiers)
-
-	apiserverConfig := p.Spec.APIServerConfig
-	shards := int32(1)
-	if p.Spec.Shards != nil && *p.Spec.Shards > 1 {
-		shards = *p.Spec.Shards
-	}
+	// Rule Files config
+	cfg = cg.AppendRuleFiles(cfg, ruleConfigMapNames, ruleSelector)
 
+	// Scrape config
 	var scrapeConfigs []yaml.MapSlice
-	for _, identifier := range sMonIdentifiers {
-		for i, ep := range sMons[identifier].Spec.Endpoints {
-			scrapeConfigs = append(scrapeConfigs,
-				cg.WithKeyVals("service_monitor", identifier).generateServiceMonitorConfig(
-					sMons[identifier],
-					ep, i,
-					apiserverConfig,
-					store,
-					shards,
-				),
-			)
-		}
-	}
-	for _, identifier := range pMonIdentifiers {
-		for i, ep := range pMons[identifier].Spec.PodMetricsEndpoints {
-			scrapeConfigs = append(scrapeConfigs,
-				cg.WithKeyVals("pod_monitor", identifier).generatePodMonitorConfig(
-					pMons[identifier], ep, i,
-					apiserverConfig,
-					store,
-					shards,
-				),
-			)
-		}
-	}
-
-	for _, identifier := range probeIdentifiers {
-		scrapeConfigs = append(scrapeConfigs,
-			cg.WithKeyVals("probe", identifier).generateProbeConfig(
-				probes[identifier],
-				apiserverConfig,
-				store,
-				shards,
-			),
-		)
+	apiserverConfig := cpf.APIServerConfig
+	shards := int32(1)
+	if cpf.Shards != nil && *cpf.Shards > 1 {
+		shards = *cpf.Shards
 	}
-
-	var addlScrapeConfigs []yaml.MapSlice
-	addlScrapeConfigs, err := cg.generateAdditionalScrapeConfigs(additionalScrapeConfigs, shards)
+	scrapeConfigs = cg.AppendServiceMonitorConfigs(scrapeConfigs, sMons, apiserverConfig, store, shards)
+	scrapeConfigs = cg.AppendPodMonitorConfigs(scrapeConfigs, pMons, apiserverConfig, store, shards)
+	scrapeConfigs = cg.AppendProbeConfigs(scrapeConfigs, probes, apiserverConfig, store, shards)
+	addlScrapeConfigs, err := cg.AppendAdditionalScrapeConfigs(cfg, additionalScrapeConfigs, shards)
```

Comment:
see my comment below on `AppendAdditionalScrapeConfigs` but I think that we want this
```suggestion
	addlScrapeConfigs, err := cg.AppendAdditionalScrapeConfigs(cfg, scrapeConfigs, shards)
```